homeassistant/components/alarm_control_panel/__init__.py

```diff
@@ -157,6 +162,80 @@ class AlarmControlPanelEntity(Entity, cached_properties=CACHED_PROPERTIES_WITH_A
     )
     _alarm_control_panel_option_default_code: str | None = None
 
+    __alarm_legacy_state: bool = False
+    __alarm_legacy_state_reported: bool = False
+
+    def __init_subclass__(cls, **kwargs: Any) -> None:
+        """Post initialisation processing."""
+        super().__init_subclass__(**kwargs)
+        if any(method in cls.__dict__ for method in ("_attr_state", "state")):
+            # Integrations should use the 'alarm_state' property instead of
+            # setting the state directly.
+            cls.__alarm_legacy_state = True
+
+    def __setattr__(self, __name: str, __value: Any) -> None:
+        """Set attribute.
+
+        Deprecation warning if settings '_attr_state' directly
+        unless already reported.
+        """
+        if __name == "_attr_state":
+            if self.__alarm_legacy_state_reported is not True:
+                self._report_deprecated_alarm_state_handling()
+            self.__alarm_legacy_state_reported = True
+        return super().__setattr__(__name, __value)
+
+    @callback
+    def add_to_platform_start(
+        self,
+        hass: HomeAssistant,
+        platform: EntityPlatform,
+        parallel_updates: asyncio.Semaphore | None,
+    ) -> None:
+        """Start adding an entity to a platform."""
+        super().add_to_platform_start(hass, platform, parallel_updates)
+        if self.__alarm_legacy_state and not self.__alarm_legacy_state_reported:
+            self._report_deprecated_alarm_state_handling()
+
+    @callback
+    def _report_deprecated_alarm_state_handling(self) -> None:
+        """Report on deprecated handling of alarm state.
+
+        Integrations should implement alarm_state instead of using state directly.
+        """
+        if self.__alarm_legacy_state_reported is True:
+            return
```

Comment:
This line isn't tested.

Reply:
Removed this. Not needed as it's checked by the callers